feat(lua): add vim.iconv

lewis6991 opened the pull request:
Identical to `vim.fn.iconv` except:

- There is no typval conversion.
- Accepts **any** Lua string, whereas typval conversion may produce a Blob type which causes `vim.fn.iconv` to fail. 
- Does not rely on `NUL` byte for string length detection.
- Can be run in fast events.
- Can be run in Lua threads (via `vim.loop.new_work`).

### Demo

Here is a screenshot of Gitsigns opening a PNG file that is managed under git.

(left `vim.iconv`, right `vim.fn.iconv`)

![image](https://user-images.githubusercontent.com/7904185/165539457-456c96c1-b89c-41a2-8c17-09c2a547216a.png)

`vim.fn.iconv`:
```lua
   for i, l in ipairs(stdout) do
      -- Have to exclude Blob types, otherwise conversion will error.
      if vim.fn.type(l) == vim.v.t_string then
        stdout[i] = vim.fn.iconv(l, encoding, 'utf-8')
      end
    end
```

`vim.iconv`:

```lua
    for i, l in ipairs(stdout) do
      stdout[i] = vim.iconv(l, encoding, 'utf-8')
    end

```

lewis6991 committed:
feat(lua): add vim.iconv

Identical to vim.fn.iconv except that there is no typval conversion and
thus accepts **any** Lua string (UTF-16/32), whereas typval conversion
may produce a Blob type which causes vim.fn.iconv to fail. Also does not
rely on NUL byte for string length detection.

Co-authored-by: Justin M. Keyes <justinkz@gmail.com>

justinmk reviewed (changes requested):
It is not sustainable for us to keep adding hand-crafted wrappers around Vimscript eval functions, just to avoid typval conversion, Blob, etc. 

> * Accepts **any** Lua string, whereas typval conversion may produce a Blob type which causes `vim.fn.iconv` to fail.
> * Does not rely on `NUL` byte for string length detection.

Why can't these things be done by the Lua <=> Vimscript bridge? Can we annotate certain functions like `f_iconv` so that the bridge knows when we want a Lua caller to get this behavior?

`vim.fn.iconv()` should just do these things implicitly.

bfredl commented on runtime/doc/lua.txt:
```diff
@@ -875,6 +875,19 @@ vim.str_byteindex({str}, {index}[, {use_utf16}])	*vim.str_byteindex()*
+        are replaced with "?".
+        The encoding names are whatever the iconv() library function
+        can accept, see ":!man 3 iconv".
+        Note that Neovim uses UTF-8 for all Unicode encodings, conversion
+        from/to UCS-2 is automatically changed to use UTF-8.  You
+        cannot use UCS-2 in a string anyway, because of the NUL bytes.
```
well this is not true with lua strings. a lua string can be UTF-16 or 32 or whatever it wants to. we should lift restrictions like this if we add this function.

lewis6991 replied:
I'm not sure what part of the code introduces this restriction. My guess is that it is done by `convert_setup` which is an alias for `convert_setup_ext` with `from_unicode_is_utf8=true` and `from_unicode_is_utf8=true`

lewis6991 commented:
> Why can't these things be done by the Lua <=> Vimscript bridge? Can we annotate certain functions like f_iconv so that the bridge knows when we want a Lua caller to get this behavior?

I've looked into this and we could annotate the function so it can run in a fast event, we should do this. I'll raise another PR at some point with all the relevant plumbing to do this for all functions that can. We'll need to somehow document this too.

The string typval conversion really doesn't work from Lua->vimscript for the main reason being typval strings don't store their length and need to rely on `NUL` termination.

> It is not sustainable for us to keep adding hand-crafted wrappers around Vimscript eval functions, just to avoid typval conversion, Blob, etc.

I agree with this in general, but I don't think this needs to be a hard rule, especially for functions which are just basic bindings onto a third-party libs (xdiff, iconv, etc).

I spent around two hours trying to fix `f_iconv` to support blob types, but didn't really get anywhere. Throwing together a basic Lua binding took about 30 minutes and only required standard Lua-C boilerplate.

justinmk commented:
> I spent around two hours trying to fix f_iconv to support blob types, but didn't really get anywhere. Throwing together a basic Lua binding took about 30 minutes and only required standard Lua-C boilerplate.

I wouldn't think we need f_iconv to support blob. If we're in Lua we can skip blob conversion and use only Lua strings?

lewis6991 commented:
I think for this specific instance, I don't know how/if we can fix the bridge plumbing to make this work. The issue here is pretty integral to typval, and I would prefer to keep any changes to typval minimal/none.

Do you have any ideas?

Otherwise this PR is the best solution.

lewis6991 commented:
>  If we're in Lua we can skip blob conversion and use only Lua strings? 

Lua strings store their length, typval strings do not. `f_iconv` only has access to typval arguments.

justinmk commented:
> Lua strings store their length, typval strings do not. `f_iconv` only has access to typval arguments.

We could add `typval_T.v_string_len`. Then annotated "Lua bridge functions" such as `f_iconv` would need to set `rettv->vval.v_string_len` before they return.

lewis6991 commented:
I had a go at sizing typval strings before I saw your edit, #18339. As you can see, it's pretty hideous and still crashes as the new size field isn't being properly initialized somewhere despite my efforts.

> Can't the Lua bridge generator code get the string length before passing it to Lua? It's not efficient, but if it works, we can think about optimizations later.

Not sure I fully understand this comment, but we need to pass the string length to typval and then pass it back to Lua, as `f_iconv` needs access to it. This is what #18339 does in addition to preventing conversion to Blob objects.

I believe messing with typval is the wrong call here. Not to mention any solution of this route will involve changes to `f_iconv` (and typval) which can't be applied back to Vim. In addition, anything in `vim.fn.*` can't (currently or ever) be run in Lua threads, which would be particularly useful for this function.

Let's leave legacy stuff as legacy and focus our efforts creating new better stuff using more robust and documented API's. It's nice the Lua<->Vimscript bridge exists, but this is really just a stop-gap rather than a first class solution. Ideas like #18306 are good because they are fairly low cost with high return.

With this PR we can effectively obsolete `f_iconv` which will only continue to exist for compat for older viml plugins, assuming any exist that use this function.

lewis6991 commented:
> It is not sustainable for us to keep adding hand-crafted wrappers around Vimscript eval functions, just to avoid typval conversion, Blob, etc.

Looping back on this. I opened #18393 to get an actual picture of where we are with Lua stdlib parity with vimscript. Turns out we've implemented ~50%, and much of the remaining 50% won't likely want/need a Lua equivalent. Additionally, #18306 will relieve the need to add a large fraction of remaining function. So, from that I don't agree with the statement.

The benefits gained and issues resolved from a Lua binding of iconv (which are entirely specific to the function at hand, with maybe the exception of the `sha256` function), fully justify its inclusion and spending more effort on the Lua-vimscript bridge isn't going to be time well spent (IMO).

justinmk commented:
> I opened #18393 to get an actual picture of where we are with Lua stdlib parity with vimscript. Turns out we've implemented ~50%, and much of the remaining 50% won't likely want/need a Lua equivalent. Additionally, #18306 will relieve the need to add a large fraction of remaining function.

That's _fantastic_ news. Thank you for digging into that. If we have a way forward that doesn't involve continuing to hand-craft Lua versions of Vimscript functions, that removes the major concern here.

> we need to pass the string length to typval and then pass it back to Lua, as `f_iconv` needs access to it.

"pass it back to Lua" is done by the bridge. Can't the bridge do a strlen() there?

> Not to mention any solution of this route will involve changes to `f_iconv` (and typval) which can't be applied back to Vim.

We must not make architecture decisions based on this. Vim-compat is "best effort" and has been for at least 2 years. It has zero weight on choices that have big maintainability and future impact on Nvim. Vim's typval is also essentially frozen since Vimscript9.

> Let's leave legacy stuff as legacy and focus our efforts creating new better stuff using more robust and documented API's. It's nice the Lua<->Vimscript bridge exists, but this is really just a stop-gap rather than a first class solution

We definitely should create better interfaces where the old one is broken. But rewriting all of the eval functions is a big waste of time. Those functions are just editor functions, there's nothing yucky about them except typval conversion. If we can make the bridge work well, then that legacy is massively valuable and avoids a huge amount of redundant work, maintenance burden, documentation burden, and user cognitive burden.

erw7 commented on src/nvim/lua/stdlib.c:
```diff
@@ -538,4 +573,8 @@ void nlua_state_add_stdlib(lua_State *const lstate, bool is_thread)
+
+  // vim.iconv
+  lua_pushcfunction(lstate, &nlua_iconv);
+  lua_setfield(lstate, -2, "iconv");
```
`iconv_string` accesses `p_ambw` and `p_emoji` internally. Therefore, it is not thread-safe. I do not agree with exposing vim.iconv to threads as it is currently implemented, although the cases where it becomes a problem may be rare.

erw7 replied:
I'm not sure what the code in question is intended to do, but I don't think it makes sense to replace characters that cannot be converted with `vim.iconv()` with the number of cell widths "`?`".

lewis6991 commented:
> If vim.fn.iconv is broken in Lua (based on the screenshot), we should throw an error when it's called from Lua.

It's only broken for strings that contain `NUL` characters, and this is already documented (more or less):

```
		Note that Vim uses UTF-8 for all Unicode encodings, conversion
		from/to UCS-2 is automatically changed to use UTF-8.  You
		cannot use UCS-2 in a string anyway, because of the NUL bytes.
```

This new version doesn't have this limitation.

There might be applications that don't need to worry about handling `NUL` bytes, so I think it would be a mistake to now suddenly make those applications error.

lewis6991 commented:
> "pass it back to Lua" is done by the bridge. Can't the bridge do a strlen() there?

`strlen` simply counts characters until it reaches `NUL`. Lua strings are allowed to contain `NUL`, so `strlen` won't work properly. The size needs to be carried the whole way through, it's information that can't be reproduced once it is gone.

> But rewriting all of the eval functions is a big waste of time. 

I dunno. Rewriting typval seems like a bigger waste of time (and less enjoyable). We could probably just implement everything we don't have a Lua version of in `vim.fn.*`, and (IMO) it would probably take less time than re-architecting typval to work in Lua's full domain. I may be underestimating, but I don't think there's a huge amount when only considering the functions we would actually want to use. Not that I think we should do this, because we simply don't need to. It's just much easier to re-implement the edge-case functions that have issues.

Also given how the Lua-C API is a real documented API, these new versions will be much more approachable to new-comers. It's how I got involved. I looked at `nvim/lua/treesitter.c`, Googled Lua-C API, and I was away.

>  Those functions are just editor functions, there's nothing yucky about them except typval conversion.

We still have don't a solution to the fact that typval doesn't have boolean values, which makes using them in Lua awkward and error-prone if you forget to compare against `0` .... which I've done many times! Not really sure how/if we can approach this in the bridge.

> If we can make the bridge work well, then that legacy is massively valuable and avoids a huge amount of redundant work, maintenance burden, documentation burden, and user cognitive burden.

I agree it is valuable, but for me it is more "we can learn from this" rather than "we can use and improve this". I would hope eventually `vim.fn.*` will become obsolete and thus deprecated (and only kept for legacy), which should remove those burdens.

One example, today I found out `bufwinid()` **only** works for the current tabpage, which I think is a bit weird. Why only the current tabpage? If we want to fix this, do we add a new option to the eval version of `bufwinid`, or can we just make something new that does and behaves exactly as we want with all the nice Lua ergonomics? I know which I'd prefer 😉 .

Anyway thanks for approving this.

lewis6991 commented:
I still need to exclude (or fix) this from running in threads.

lewis6991 commented:
Note I've needed to add `#if defined(HAVE_ICONV)` guards, since `iconv` is not a required build dependency.

erw7 commented:
Crash(segmentation fault) with the following steps:

```
nvim --clean
:lua vim.iconv('a', 'foo', 'bar')
```

I think it is necessary to do the same as the following process of `f_iconv`, or call `luaL_error()` if `ret` is `NULL`. I think `iconv()` in vim script is insufficient for some purposes, such as not being able to know if there are characters that cannot be converted. It might be better to change it to `vim.iconv({string}, {from}, {to} [, {opt}])` and allow `fail` and `replacement character` to be specified by `opt`.

https://github.com/neovim/neovim/blob/b3453ea4a6b6518cccd6922fbcaa64a0aa47c743/src/nvim/eval/funcs.c#L4620-L4622

lewis6991 commented:
@erw7: 
* Fixed the crash
* Added an `opts` argument with a `fail` field
* Didn't add `replacement text` as that is less-trivial and can be done in a future PR
* Added a few tests
* Also changed the return to be `nil` when conversion fails as that is more "Lua".

justinmk commented:
asan failure:

```
[ RUN      ] lua stdlib vim.iconv can validate arguments: 29.55 ms OK
==================== File /home/runner/work/neovim/neovim/build/log/ubsan.36031 ====================
= 
= =================================================================
= ==36031==ERROR: LeakSanitizer: detected memory leaks
= 
= Direct leak of 11 byte(s) in 1 object(s) allocated from:
=     #0 0x4de7ad in malloc ??:?
=     #1 0x1267e92 in try_malloc /home/runner/work/neovim/neovim/src/nvim/memory.c:74:15
=     #2 0x126808c in xmalloc /home/runner/work/neovim/neovim/src/nvim/memory.c:108:15
=     #3 0x1208b6b in string_convert_ext /home/runner/work/neovim/neovim/src/nvim/mbyte.c:2547:14
=     #4 0x1200148 in string_convert /home/runner/work/neovim/neovim/src/nvim/mbyte.c:2519:10
=     #5 0x10a9c53 in nlua_iconv /home/runner/work/neovim/neovim/src/nvim/lua/stdlib.c:526:23
=     #6 0x1f2b8d7 in luaD_precall /home/runner/nvim-deps/build/src/lua/src/ldo.c:320:10
=     #7 0x1f3a1f3 in luaV_execute /home/runner/nvim-deps/build/src/lua/src/lvm.c:612:17
=     #8 0x1f2bb8a in luaD_call /home/runner/nvim-deps/build/src/lua/src/ldo.c:378:5
=     #9 0x1f28027 in f_call /home/runner/nvim-deps/build/src/lua/src/lapi.c:800:3
=     #10 0x1f2ab94 in luaD_rawrunprotected /home/runner/nvim-deps/build/src/lua/src/ldo.c:116:3
=     #11 0x1f2bf94 in luaD_pcall /home/runner/nvim-deps/build/src/lua/src/ldo.c:464:12
=     #12 0x1f280e9 in lua_pcall /home/runner/nvim-deps/build/src/lua/src/lapi.c:821:12
=     #13 0x1093729 in nlua_pcall /home/runner/work/neovim/neovim/src/nvim/lua/executor.c:130:16
=     #14 0x10959f3 in nlua_exec /home/runner/work/neovim/neovim/src/nvim/lua/executor.c:1322:7
=     #15 0x739f78 in nvim_exec_lua /home/runner/work/neovim/neovim/src/nvim/api/vim.c:459:10
=     #16 0x69bf71 in handle_nvim_exec_lua

lewis6991 commented:
Removed the `opts` field, instead just check for `nil` on the return value to know if conversion failed.
